src/apps/polymer/vIBCEscrow.sol

```diff
@@ -0,0 +1,209 @@
+        }
+    }
+
+    function onCloseIbcChannel(bytes32 channelId, string calldata, bytes32) external virtual onlyIbcDispatcher {
+        // logic to determin if the channel should be closed
+        bool channelFound = false;
+        for (uint256 i = 0; i < connectedChannels.length; i++) {
+            if (connectedChannels[i] == channelId) {
+                delete connectedChannels[i];
+                channelFound = true;
+                break;
+            }
+        }
+        if (!channelFound) revert ChannelNotFound();
+    }
+
+    //--- IBC Packet Callbacks ---//
+
+    // packet.srcPortAddr is the IncentivizedPolymerEscrow address on the source chain.
+    // packet.destPortAddr is the address of this contract.
+    // channelId: the channel id from the running chain's perspective, which can be used to identify the counterparty chain.
+    function onRecvPacket(IbcPacket calldata packet)
+        external override
+        onlyIbcDispatcher
+        returns (AckPacket memory)
+    {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        // Collect the implementation identifier we added. Remember, this is trusted IFF packet.src.channelId is trusted.
+        // sourceImplementationIdentifier has already been defined by the channel on channel creation.
+        bytes memory sourceImplementationIdentifier = packet.data[POLYMER_SENDER_IDENTIFIER_START:POLYMER_SENDER_IDENTIFIER_END];
+
+        bytes memory receiveAck = _handleMessage(
+            packet.src.channelId,
+            sourceImplementationIdentifier,
+            packet.data[POLYMER_PACKAGE_PAYLOAD_START: ],
+            feeRecipitent,
+            gasLimit
+        );
+
+        // Send ack:
+        return AckPacket({success: true, data: bytes.concat(ADDRESS_THIS, receiveAck)});
+    }
+
+    function onAcknowledgementPacket(
+        IbcPacket calldata packet,
+        AckPacket calldata ack
+    )
+        external override
+        onlyIbcDispatcher
+    {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        // Collect the implementation identifier we added. Remember, this is trusted IFF packet.src.channelId is trusted.
+        bytes memory destinationImplementationIdentifier = ack.data[POLYMER_SENDER_IDENTIFIER_START:POLYMER_SENDER_IDENTIFIER_END];
+
+        // Get the payload by removing the implementation identifier.
+        bytes calldata rawMessage = ack.data[POLYMER_PACKAGE_PAYLOAD_START:];
+
+        // Set a verificaiton context so we can recover the ack.
+        isVerifiedMessageHash[keccak256(rawMessage)] = VerifiedMessageHashContext({
+            chainIdentifier: packet.src.channelId,
+            implementationIdentifier: destinationImplementationIdentifier
+        });
+        _handleAck(packet.src.channelId, destinationImplementationIdentifier, rawMessage, feeRecipitent, gasLimit);
```

Comment:
If `_handleAck` reverts, `isVerifiedMessaageHash` state wouldn't be mutated. So how to recover the ack? this is related to the question above too

Reply:
That is a good question. What is the execution strictness when delivered from Polymer?
Is it required for applications that the `on....Packet` functions never revert?
So we need to call `_handleAck` within a try statement?
What happens if `onAcknowledgementPacket` reverts from `Polymer's perspective?

Generally, `_handleAck` should never revert if the relayer provides enough gas for the application. It is only if the relayer cheaps out the function can revert.

Reply:
On Polymer side, all packet callback functions `on...Packet` are effectively try-catched, eg. `onAcknowledgementPacket`[(bool success, bytes memory data) = _callIfContract(](https://github.com/open-ibc/vibc-core-smart-contracts/blob/893890793a5cb2d5d534604b2d9aec67b4fb0c28/contracts/core/Dispatcher.sol#L351). So it's ok if IBC-dApps revert in any callback functions. 

In your case, since retry is handled by `recoverAck`, I'd recommend you wrap `_handleAck` in try-catch, so to record acks even when `_handleAck` reverts due to lack of gas.

Reply:
The try-catch implementation in the dispatcher is vulnerable to gas denial of service for large gas spenders. This is because max 63/64 of the gas is forwarded. There should be an gas check, otherwise there is a soft-max gas limit for Polymer applications.

Something like this:
https://github.com/catalystdao/GeneralisedIncentives/blob/e410087b6faca4ce737b50a74f276e27ce7874ad/src/IncentivizedMessageEscrow.sol#L585-L587

This is not a catch-all solution since reverting by spending all gas will block from the applications side. But at least that will be the applications fault.

Looking further through your implementation, this is not an immediate using it for our testnet purposes.

Our contracts should not revert when packages are properly relayed and as a result, the try-catch in itself is not what we worry about. Furthermore, there is still a way to process the acks since acks are not written (on Polymer) until they execute without failure.